### ISSUE TITLE
Aggregator example should not create new nodes from existing nodes

### ISSUE THREAD

**Issue by tiziano88** [project-oak/oak#1246]
https://github.com/project-oak/oak/blob/dc23f27ad67f1707a8d92fb126fed932a80f4599/examples/aggregator/module/rust/src/lib.rs#L108-L111

Once we start labelling created nodes with more sensible labels, the main programming pattern will be for all nodes and channels to be created directly by the initial node of an application, while "worker" nodes just receive handles to pre-existing objects.

For the aggregator example, this means creating the gRPC client pseudo-node as part of the `oak_main` entrypoint:

https://github.com/project-oak/oak/blob/dc23f27ad67f1707a8d92fb126fed932a80f4599/examples/aggregator/module/rust/src/lib.rs#L166-L182

and then pass its handle to the gRPC worker.

The complication is that the gRPC worker currently only accepts gRPC invocation messages as input, so it needs to be modified so that it can accept a new type of message which is essentially an enum of either an "init" message (containing the initial handles and any other configuration), or the actual gRPC message to process.

**Comment by ipetr0v** [comment]
I remember that I have reviewed a commit that was adding an "init" message, but I don't remember where it was, and have it been added yet.

**Comment by ipetr0v** [comment]
Found it:
https://github.com/project-oak/oak/blob/7f33be4ff8f1a0949514643e36ad9f1a2253141c/sdk/rust/oak/src/grpc/server.rs#L21-L23

But it's a different "init" message.

**Comment by ipetr0v** [comment]
Also, why should it be an enum of {Init, gRPC_invocation}?
Could it just read the first init message and then pass handle to the gRPC dispatcher?

**Comment by tiziano88** [comment]
I don't think it's a good idea to rely on the order of the messages to figure out what they should do. We should use strongly typed messages instead, and each node should handle only one type of messages ever, rather than switch between different types.

So for this case, I think the implementation of the aggregator node should accept a `oneof` between an init message, and a gRPC invocation. If the gRPC invocation variant is received, then it should be dispatched using the code generated by prost.

cc @conradgrobler as we chatted about this earlier.

**Comment by ipetr0v** [comment]
We are also loading `ConfigMap` the same way (as a first message in the initial handle):
https://github.com/project-oak/oak/blob/7b89fbdada697230672eae48b10fbe7904a33b2f/examples/trusted_information_retrieval/module_0/rust/src/lib.rs#L76-L79

**Comment by tiziano88** [comment]
That handle is not used for anything else afterwards though

**Comment by tiziano88** [comment]
Basically the "input type" of the main node is `ConfigMap`.

The input type of the gRPC node should be a message which is either an `Init` variant or a `GrpcInvocation` variant, but it's just a single enum / oneof type, with two variants.

WDYT @ipetr0v ?

**Comment by ipetr0v** [comment]
Agree
Currently we have a `handle_command` function:
https://github.com/project-oak/oak/blob/7b89fbdada697230672eae48b10fbe7904a33b2f/sdk/rust/oak/src/grpc/mod.rs#L149-L159
That is being run by `run_event_loop`:
https://github.com/project-oak/oak/blob/7b89fbdada697230672eae48b10fbe7904a33b2f/sdk/rust/oak/src/lib.rs#L455-L458
https://github.com/project-oak/oak/blob/7b89fbdada697230672eae48b10fbe7904a33b2f/sdk/rust/oak/src/lib.rs#L508-L513

I think it may be changed on that level, since there is nothing much to do about it inside the application module.
Or we can just add a match in the `entrypoint`:
https://github.com/project-oak/oak/blob/7b89fbdada697230672eae48b10fbe7904a33b2f/examples/aggregator/module/rust/src/lib.rs#L160-L164

**Comment by ipetr0v** [comment]
Currently we generate:
```rust
pub trait Aggregator { 
    fn submit_sample ( & mut self , req : Sample ) -> :: oak :: grpc :: Result < ( ) > ;
}
```
and use it as a generic parameter for:
```rust
impl < T : Aggregator > :: oak :: grpc :: ServerNode for AggregatorDispatcher < T > {
    fn invoke ( & mut self , method : & str , req : & [ u8 ] , writer : :: oak :: grpc :: ChannelResponseWriter ) {
        match method {
            "/oak.examples.aggregator.Aggregator/SubmitSample" => {
                ::oak::grpc::handle_req_rsp(|r| (self.0).submit_sample(r), req, writer)
            },
            _ => {
                ::log::error!("unknown method name: {}", method);
            }
        }
    }
}
```

I think we can add an `init` function, that takes a list of handles as a parameter, and should be implemented in the application `module`:
```rust
pub trait Aggregator {
    fn init(&mut self, nodes: Node<Handle>);
    fn submit_sample ( & mut self , req : Sample ) -> :: oak :: grpc :: Result < ( ) > ;
}
```

**Comment by tiziano88** [comment]
Please don't pass unstructured handles like that, it is the equivalent of passing "bytes" to a method instead of properly typed values. Instead, define a protobuf message with fields of type handle, with appropriate names.

**Comment by tiziano88** [comment]
Also, I think it would be nicer if the init message was received separatedly, and used to initialise the struct. Another issue with the proposed approach is that if the protobuf service definition has a method called `init`, it would clash with it.

**Comment by tiziano88** [comment]
Any update on this? happy to keep discussing if it is helpful :slightly_smiling_face:

**Comment by ipetr0v** [comment]
And starter receiving the following errors:
```
error[E0277]: the trait bound `proto::oak::encap::GrpcResponse: handle::HandleVisit` is not satisfied
  --> /opt/my-project/sdk/target/debug/build/oak-c66b74b6dc5405b8/out/oak.invocation.rs:6:5
   |
6  |     pub response_sender: ::core::option::Option<crate::io::Sender<super::encap::GrpcResponse>>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `handle::HandleVisit` is not implemented for `proto::oak::encap::GrpcResponse`
   | 
  ::: rust/oak/src/io/sender.rs:30:22
   |
30 | pub struct Sender<T: Encodable> {
   |                      --------- required by this bound in `io::sender::Sender`
   |
   = note: required because of the requirements on the impl of `io::encodable::Encodable` for `proto::oak::encap::GrpcResponse`
```
Looks like I need to update the generator code.

**Comment by tiziano88** [comment]
Is this blocked on #1249? are we in a deadlock? :)
